File: python/yacht/yacht.py
"""
This exercise stub and the test suite contain several enumerated constants.

Since Python 2 does not have the enum module, the idiomatic way to write
enumerated constants has traditionally been a NAME assigned to an arbitrary,
but unique value. An integer is traditionally used because it’s memory
efficient.
It is a common practice to export both constants and functions that work with
those constants (ex. the constants in the os, subprocess and re modules).

You can learn more here: https://en.wikipedia.org/wiki/Enumerated_type
"""


# Score categories.
# Change the values as you see fit.
(
    YACHT,
    ONES,
    TWOS,
    THREES,
    FOURS,
    FIVES,
    SIXES,
    FULL_HOUSE,
    FOUR_OF_A_KIND,
    LITTLE_STRAIGHT,
    BIG_STRAIGHT,
    CHOICE,
) = range(0, 12)


def score(dice, category):
    methods = {
        0: lambda dice: 50 if total_sides(dice).get(5, 0) > 0 else 0,
        7: lambda dice: sum(dice) if is_full_house(dice) else 0,
        8: lambda dice: max(dice) * 4 if is_four_a_kind(dice) else 0,
        9: lambda dice: 30 if sorted(dice) == [1, 2, 3, 4, 5] else 0,
        10: lambda dice: 30 if sorted(dice) == [2, 3, 4, 5, 6] else 0,
        11: lambda dice: sum(dice),
    }

    def calc_score(dice, side, score):
        return dice.count(side) * score

    def total_sides(dice):
        return {dice.count(side): side for side in range(1, 7)}

    def is_full_house(dice):
        sides = total_sides(dice)
        return sides.get(2, 0) > 0 and sides.get(3, 0) > 0

    def is_four_a_kind(dice):
        sides = total_sides(dice)
        return sides.get(4, 0) > 0 or sides.get(5, 0) > 0

    return (
        calc_score(dice, side=category, score=category)
        if category in range(1, 7)
        else methods[category](dice)
    )

File: go/yacht/yacht.go
package yacht

import (
	"slices"
)

const (
	YACHT           = "yacht"
	ONES            = "ones"
	TWOS            = "twos"
	THREES          = "threes"
	FOURS           = "fours"
	FIVES           = "fives"
	SIXES           = "sixes"
	FULL_HOUSE      = "full house"
	FOUR_OF_A_KIND  = "four of a kind"
	LITTLE_STRAIGHT = "little straight"
	BIG_STRAIGHT    = "big straight"
	CHOICE          = "choice"
)

var scores map[string]int = map[string]int{
	ONES:   1,
	TWOS:   2,
	THREES: 3,
	FOURS:  4,
	FIVES:  5,
	SIXES:  6,
}
var categories []string = []string{
	ONES,
	TWOS,
	THREES,
	FOURS,
	FIVES,
	SIXES,
}

func Score(dice []int, category string) int {
	fns := make(map[string]func([]int) int, 6)

	fns[YACHT] = func(dice []int) int {
		total := totalSides(dice)
		if total[5] > 0 {
			return 50
		}
		return 0
	}

	fns[FULL_HOUSE] = func(dice []int) int {
		if isFullHouse(dice) {
			return sum(dice)
		}
		return 0
	}

	fns[FOUR_OF_A_KIND] = func(dice []int) int {
		if isFourAKind(dice) {
			return slices.Max(dice) * 4
		}
		return 0
	}

	fns[LITTLE_STRAIGHT] = func(dice []int) int {
		slices.Sort(dice)
		if slices.Compare(dice, []int{1, 2, 3, 4, 5}) == 0 {
			return 30
		}
		return 0
	}

	fns[BIG_STRAIGHT] = func(dice []int) int {
		slices.Sort(dice)
		if slices.Compare(dice, []int{2, 3, 4, 5, 6}) == 0 {
			return 30
		}
		return 0
	}

	fns[CHOICE] = func(dice []int) int {
		return sum(dice)
	}

	if slices.Index(categories, category) != -1 {
		return calcScore(dice, scores[category], scores[category])
	}

	return fns[category](dice)
}

func calcScore(dice []int, side int, score int) int {
	total := countSides(dice)
	return total[side] * score
}

func sum(dice []int) int {
	var r int
	for _, side := range dice {
		r += side
	}
	return r
}

func countSides(dice []int) map[int]int {
	var r map[int]int = map[int]int{1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
	for _, side := range dice {
		r[side] += 1
	}
	return r
}

func totalSides(dice []int) map[int]int {
	totals := make(map[int]int, 0)
	for side, count := range countSides(dice) {
		totals[count] = side
	}
	return totals
}

func isFourAKind(dice []int) bool {
	total := totalSides(dice)
	return total[4] > 0 || total[5] > 0
}

func isFullHouse(dice []int) bool {
	total := totalSides(dice)
	return total[2] > 0 && total[3] > 0
}
